lib/consensus/raft/consensus.go

```diff
@@ -0,0 +1,563 @@
+		// we're done here.
+
+		//ok, err := cc.RedirectToLeader(rpcOp, redirectArg, struct{}{})
+		//if err != nil || ok {
+		//	return err
+		//}
+
+		// Being here means we are the LEADER. We can commit.
+
+		// now commit the changes to our state
+		//cc.shutdownLock.RLock() // do not shut down while committing
+		_, finalErr = cc.consensus.CommitOp(op)
+		//cc.shutdownLock.RUnlock()
+		if finalErr != nil {
+			goto RETRY
+		}
+
+	RETRY:
+		time.Sleep(cc.config.CommitRetryDelay)
+	}
+	return finalErr
+}
+
+// AddPeer adds a new peer to participate in this consensus. It will
+// forward the operation to the leader if this is not it.
+func (cc *Consensus) AddPeer(ctx context.Context, pid peer.ID) error {
+	//ctx, span := trace.StartSpan(ctx, "consensus/AddPeer")
+	//defer span.End()
+
+	var finalErr error
+	for i := 0; i <= cc.config.CommitRetries; i++ {
+		logger.Debugf("attempt #%d: AddPeer %s", i, pid.Pretty())
+		if finalErr != nil {
+			logger.Errorf("retrying to add peer. Attempt #%d failed: %s", i, finalErr)
+		}
+		ok, err := cc.RedirectToLeader("AddPeer", pid, struct{}{})
+		if err != nil || ok {
+			return err
+		}
+		// Being here means we are the leader and can commit
+		//cc.shutdownLock.RLock() // do not shutdown while committing
+		//finalErr = cc.raft.AddPeer(ctx, peer.Encode(pid))
+		finalErr = cc.raft.AddPeer(ctx, pid)
+
+		//cc.shutdownLock.RUnlock()
+		if finalErr != nil {
+			time.Sleep(cc.config.CommitRetryDelay)
+			continue
+		}
+		logger.Infof("peer added to Raft: %s", pid.Pretty())
+		break
+	}
+	return finalErr
+}
+
+// RmPeer removes a peer from this consensus. It will
+// forward the operation to the leader if this is not it.
+func (cc *Consensus) RmPeer(ctx context.Context, pid peer.ID) error {
+	//ctx, span := trace.StartSpan(ctx, "consensus/RmPeer")
+	//defer span.End()
+
+	var finalErr error
+	for i := 0; i <= cc.config.CommitRetries; i++ {
+		logger.Debugf("attempt #%d: RmPeer %s", i, pid.Pretty())
+		if finalErr != nil {
+			logger.Errorf("retrying to remove peer. Attempt #%d failed: %s", i, finalErr)
+		}
+		ok, err := cc.RedirectToLeader("RmPeer", pid, struct{}{})
+		if err != nil || ok {
+			return err
+		}
+		// Being here means we are the leader and can commit
+		//cc.shutdownLock.RLock() // do not shutdown while committing
+		finalErr = cc.raft.RemovePeer(ctx, peer.Encode(pid))
+		//cc.shutdownLock.RUnlock()
+		if finalErr != nil {
+			time.Sleep(cc.config.CommitRetryDelay)
+			continue
+		}
+		logger.Infof("peer removed from Raft: %s", pid.Pretty())
+		break
+	}
+	return finalErr
+}
+
+// RaftState retrieves the current consensus RaftState. It may error if no RaftState has
+// been agreed upon or the state is not consistent. The returned RaftState is the
+// last agreed-upon RaftState known by this node. No writes are allowed, as all
+// writes to the shared state should happen through the Consensus component
+// methods.
+func (cc *Consensus) State(ctx context.Context) (consensus.State, error) {
```

Comment:
Don't we want to return `RaftState` instead of `consensus.State` here?

Reply:
consensus.State is an interface implemented by RaftState